include/fiction/algorithms/simulation/sidb/quickexact.hpp

```diff
@@ -0,0 +1,506 @@
+     * run even if base_num = 2 was set in the physical parameters.
+     */
+    ON,
+    /**
+     * The base number from the physical parameter is used for the simulation.
+     */
+    OFF
+};
+/**
+ * This struct stores the parameters for the `quickexact` algorithm.
+ */
+template <typename Lyt>
+struct quickexact_params
+{
+    /**
+     * All parameters for physical SiDB simulations.
+     */
+    sidb_simulation_parameters physical_parameters{};
+    /**
+     * If ON, QuickExact checks before which base number is required for the simulation, i.e., whether 3-state is
+     * necessary or 2-state simulation is sufficient.
+     */
+    automatic_base_number_detection base_number_detection = automatic_base_number_detection::ON;
+    /**
+     * Local external electrostatic potentials (e.g locally applied electrodes).
+     */
+    std::unordered_map<typename Lyt::cell, double> local_external_potential = {};
+    /**
+     * Global external electrostatic potential. Value is applied on each cell in the layout.
+     */
+    double global_potential = 0;
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class quickexact_impl
+{
+  public:
+    quickexact_impl(Lyt& lyt, const quickexact_params<Lyt>& parameter) :
+            layout{lyt},
+            charge_lyt{lyt, parameter.physical_parameters, sidb_charge_state::NEGATIVE},
+            params{parameter}
+    {}
+
+    sidb_simulation_result<Lyt> run()
+    {
+        result.algorithm_name      = "quickexact";
+        result.physical_parameters = params.physical_parameters;
+
+        mockturtle::stopwatch<>::duration time_counter{};
+        {
+            const mockturtle::stopwatch stop{time_counter};
+
+            initialize_charge_layout();
+
+            //  Determine all SiDBs that have to be negatively charged to fulfill the population stability. This is an
+            //  efficient way to prune the search space by 2^k with k being the number of detected negatively-charged
+            //  SiDBs.
+
+            //  Determine if three state simulation (i.e. positively charged SiDBs can occur) is required.
+            const bool three_state_simulation_required =
+                params.base_number_detection == automatic_base_number_detection::ON &&
+                charge_lyt.three_state_simulation_required();
+
+            // If layout has at least two SiDBs, the code inside this if-statement is executed.
+            if (number_of_sidbs > 1)
+            {
+                generate_layout_without_negative_sidbs();
+            }
+
+            if (!all_sidbs_in_lyt_without_negative_detected_ones.empty() && number_of_sidbs > 1)
+            {
+                // The first cell from all_sidbs_in_lyt_without_negative_detected_ones is chosen as the dependent cell
+                // to initialize the layout (detected negatively-charged SiDBs were erased in the step before).
```

Comment:
What do you mean by "the step before"? What is a "step" in your algorithm?

Reply:
if statement above.